addons/account_tax_python/models/account_tax.py

```diff
@@ -29,33 +28,68 @@ class AccountTaxPython(models.Model):
             ":param product: product.product recordset singleton or None\n"
             ":param partner: res.partner recordset singleton or None")
 
-    def _compute_amount(self, base_amount, price_unit, quantity=1.0, product=None, partner=None, fixed_multiplicator=1):
-        self.ensure_one()
-        if product and product._name == 'product.template':
-            product = product.product_variant_id
-        if self.amount_type == 'code':
-            company = self.env.company
-            localdict = {'base_amount': base_amount, 'price_unit':price_unit, 'quantity': quantity, 'product':product, 'partner':partner, 'company': company}
+    def _prepare_dict_for_taxes_computation(self):
+        # EXTENDS 'account'
+        tax_values = super()._prepare_dict_for_taxes_computation()
+        tax_values.update({
+            'python_compute': self.python_compute,
+            'python_applicable': self.python_applicable,
+        })
+        return tax_values
+
+    @api.model
+    def _ascending_process_fixed_taxes_batch(self, batch):
+        # EXTENDS 'account'
+        super()._ascending_process_fixed_taxes_batch(batch)
+
+        if batch['amount_type'] == 'code':
+            batch['computed'] = 'tax'
+
+    @api.model
+    def _descending_process_price_included_taxes_batch(self, batch):
+        # EXTENDS 'account'
+        super()._descending_process_price_included_taxes_batch(batch)
+
+        if batch['price_include'] and batch['amount_type'] == 'code':
+            batch['computed'] = True
+
+    @api.model
+    def _ascending_process_taxes_batch(self, batch):
+        # EXTENDS 'account'
+        super()._ascending_process_taxes_batch(batch)
+
+        if not batch['price_include'] and batch['amount_type'] == 'code':
+            batch['computed'] = True
+
+    @api.model
+    def _eval_tax_amount(self, tax_values, evaluation_context):
+        # EXTENDS 'account'
+        amount_type = tax_values['amount_type']
+        if amount_type == 'code':
+            raw_base = (evaluation_context['quantity'] * evaluation_context['price_unit']) + evaluation_context['extra_base']
+            local_dict = {**evaluation_context, 'base_amount': raw_base}
+            try:
+                safe_eval(tax_values['python_applicable'], local_dict, mode="exec", nocopy=True)
```

Comment:
@odoo/rd-security Moved safe_eval to another place

Reply:
@xmo-odoo
`tax_values` comes from `_prepare_dict_for_taxes_computation` in the same file. It's basically the values of the `account.tax`.
`evaluation_context` comes from `_eval_taxes_computation_prepare_context` in account/models/account_tax.py.
The user doesn't have any access to those methods. Everything is computed from values stored in an odoo model.

Reply:
Can we make some sanity check like ensuring every record in the context is numerical or instance of BaseModel?

Reply:
There is absolutely no record in those dictionaries because those methods are also defined for the js-side. However, account_python_tax doesn't work js-side and for backward compatibility (I'll backport this PR in 17.0), I need to give the `product` record in the safe-eval. I have a plan to remove this module later in master but for now, I need to keep the previous behavior :x

Reply:
> However, account_python_tax doesn't work js-side and for backward compatibility (I'll backport this PR in 17.0), I need to give the product record in the safe-eval.

I don't get it. Let's do it correctly directly and get rid of this product as soon as we can if it's not a functionnal requirement. We will see for a minimal patch in 17.0 later but make it right in master please.

Reply:
@mart-e What are you talking about? I can't "get rid" of the python module right now. If this module introduce a big security hole because it gives access to the product in a safe_eval, then we have an issue since at least v9 but that's not a reason to block this task. I'm conflicting everyday regarding the size of the diff. Let's consider it's a separated topic (impact all stable versions then)